Semi-automatic scripts on the last MCP version

Elisa-Visentin opened the pull request:
Update of #170 (which was too old wrt master). Comments in the old PR will be considered here

Authors: Elisa, Federico, Joanna, Julian, Raniere

Elisa-Visentin commented:
The last committed scripts can be used to create an LST-run-wise database. Here its structure
![image](https://github.com/cta-observatory/magic-cta-pipe/assets/121040436/131e90b3-6389-4782-8282-283e96d8e91c)
nsb column hosts NSB level, lstchain_0.9 and lstchain_0.10 columns will host a bool value (i.e., True if version 0.9/0.10 for this run exists. STILL TO BE FILLED), error_code is nan in case of good runs, 000 in case NSB cannot be evaluated, 001 in case NSB > 3.0
Suggestions? What about this structure?

jsitarek commented on magicctapipe/scripts/lst1_magic/README.md:
````diff
@@ -0,0 +1,226 @@
+2) Once connected to the IT Container, install magic-cta-pipe (e.g. in your home directory in the IT Container) with the following commands (if you have mamba installed, we recommend you to use it instead of conda. The installation process will be much faster.):
+
+```
+git clone -b Torino_auto_MCP https://github.com/cta-observatory/magic-cta-pipe.git
````
this will have to be updated before the PR is merged, now as a draft it is ok

Elisa-Visentin replied:
done

jsitarek commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,226 @@
+
+## Analysis
+
+During the analysis, some files are automatically produced by the scripts and are saved on your working directory. These files are necessary for the subsequent steps in the analysis chain. It is furthermore mandatory that you always launch the scripts from the same working directory since the output files stored there can be correctly read as input files at the subsequent analysis steps.
```
I think it would be good to be able to specify such a temporary directory. If those are e.g. log files, it is a good idea to keep them close to the place where the output files are.

Elisa-Visentin replied:
These files are: list of runs to be processed, list of source names, sh files and, for us (we heave to update the LST database) the txt files used to fill the nsb column. Log files are saved in a 'logs' folder in each output directory

Elisa-Visentin replied:
EDIT: now slurm output files are redirected to output folders with the logs. And maybe we could delete the sh files after they have been completed (with exit code 0) to 'clean' the working directory and leave only list of runs and sources there

jsitarek replied:
deleting the .sh files would be nice, but it is not easy, because they are the files that are actually executed by the cluster so you can remove them only after the job is done (and the script probably cannot remove itself, in particular if it is run in an array). In the tests that I did I also saw how many files are those (at least in DL1 step) and it is fine to clean them up by hand afterwards, so I close the comment

jsitarek commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,226 @@
+directories:
+    workspace_dir : "/fefs/aswg/workspace/yourname/yourprojectname/"  # Output directory where all the data products will be saved.
+    # MC paths below are ignored if you set NSB_matching = true.
+    MC_gammas     : "/fefs/aswg/data/mc/DL0/LSTProd2/TestDataset/sim_telarray"  # set to "" if you don't want to process these Monte Carlo simulations.
```
even we use preprocessed MCs, we will need at some point the RF files for DL2 data files and the IRF files for producing DL3 files. 
If those are going to be used from the general MC table, we will need to specify some MC-related settings: analysis period and processing version, unless the scripts will figure them out themselfs (important if the date range requested involves multiple periods)

Elisa-Visentin replied:
As for the analysis period, if you mean the MAGIC obs period (e.g., ST0317) the scripts will (?) estimate the right period for each observation (based on dates) and search for the MC of the right period for each run, so that you can process in the same analysis data collected over >2 years (indeed, in the previous version we created one folder per period since MAGIC DL1 and we matched data-MC periods based on the directories names). As for MC processing version, now there is only one version and it is hard-coded in the paths  in the DL1 to DL2 and DL2 to DL3 automatic scripts (still private scripts, but they exist ;) ), but we can for sure add an entry in the config file for this version

jsitarek replied:
ok, since this is for DL2 ==> DL3, and those scripts are not a part of this PR, I'm closing this for the moment (and will check it again once we reach to DL3)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+    lines = [
+        f"mc_tel_ids:\n    LST-1: {ids[0]}\n    LST-2: {ids[1]}\n    LST-3: {ids[2]}\n    LST-4: {ids[3]}\n    MAGIC-I: {ids[4]}\n    MAGIC-II: {ids[5]}\n\n",
+        'event_coincidence:\n    timestamp_type_lst: "dragon_time"  # select "dragon_time", "tib_time" or "ucts_time"\n    pre_offset_search: true\n    n_pre_offset_search_events: 100\n    window_half_width: "300 ns"\n',
+        '    time_offset:\n        start: "-10 us"\n        stop: "0 us"\n',
```
I do not like this approach much, because it hardcodes the settings inside the script, and if we change them in the general config file they are not propagated. How about instead you pass the (path to) general config file instead, and extract from it mc_tel_ids:  and event_coincidence: ?

Elisa-Visentin replied:
Yes for sure. We already tried to fix this in the setting_up script: we can use the same method also here and in stereo, if the method is fine for you

aleberti replied:
I agree with Julian here, the general config has all the information needed already.

Elisa-Visentin replied:
Fixed (?)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+        Path to the working directory
+    LST_runs : matrix of strings
+        This matrix is imported from config_general.yaml and tells the function where to find the LST data and link them to our working directory
+    nsb : int
```
nsb is not an int (can be e.g. 0.5)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+    target_dir : str
+        Path to the working directory
+    LST_runs : matrix of strings
+        This matrix is imported from config_general.yaml and tells the function where to find the LST data and link them to our working directory
```
I do not see this matrix in config_general.yaml

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+        os.path.basename(x) for x in glob.glob(f"{target_dir}/v{__version__}/DL1/*")
+    ]
+
+    if (len(LST_runs) == 2) and (len(LST_runs[0]) == 10):
```
this looks strange

Elisa-Visentin replied:
Do you mean line 88? In case there is only one entry (i.e., one line) in the MAGIC_runs.txt or LST_runs.txt, LST_runs won't be a set of [date,run] couples (i.e., list of lists  [[xx,yy],[xx,yy]....]) but will be read as a list of two elements ([date,run]) and will thus split run and date instead of using them as a couple; this if clause fixes this issue by 'adding two brackets' around the list in the latter case. This way, we will always deal with [[...]....] objects. But any suggestion here is welcome

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                for x in glob.glob(f"{MAGIC_DL1_dir}/Merged/Merged_*")
+            ]
+            for d in dates:
+                Y_M = int(d.split("_")[1])
```
how about:
```Y_M, M_M, D_M = [int(x) for x in d.split("_")[1:]]```

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+
+                delta = timedelta(days=1)
+                for i in LST_runs:
+                    Y_L = i[0].split("_")[0]
```
same here (if there are only 3 fields, otherwise you can add also ...[0:3]: 
```Y_L, M_L, D_L = i[0].split("_")```

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                                f"{coincidence_DL1_dir}/DL1Coincident/{p}/NSB{nsb}"
+                            )
+
+                        lstObsDir = (
```
if i[0] is only a date you can do:
```lstObsDir = i[0].replace('_', '')```

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                                f"{coincidence_DL1_dir}/DL1Coincident/{p}/NSB{nsb}/{lstObsDir}/logs"
+                            )
+
+                        outputdir = f"{coincidence_DL1_dir}/DL1Coincident/{p}/NSB{nsb}/{lstObsDir}"
```
instead of the two ifs above you can use immediately after this line:
```os.makedirs(outputdir+"/logs")```
this one will automatically create the parent directories and will also not complain if the directory already exists

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                        list_of_subruns = np.sort(
+                            glob.glob(f"{inputdir}/dl1*Run*{i[1]}*.*.h5")
+                        )
+                        if os.path.exists(f"{outputdir}/logs/list_LST"):
```
I think you do not need an if in here. If you simply open the file with 'a' mode, even if it does not exist it will be created, this will simplify the code

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+
+
+def linking_bash_lst(
+    target_dir, LST_runs, nsb, date, source, LST_version, env_name, NSB_match
```
you should pass also queueing system type to this function because you produce a bash file following slurm style

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                            "#SBATCH -p short\n",
+                            f'#SBATCH -J {target_dir.split("/")[-2:][1]}_coincidence_{nsb}\n',
+                            f"#SBATCH --array=0-{process_size}\n",
+                            "#SBATCH --mem=30g\n",
```
30 GB is a lot. I normally run it with 4 GB, however for some runs it is not enough. We could keep e.g. 8 GB as a default. A more clever system would run it first with size of the memory that is fine for most of the runs and if it fails the jot is resubmitted with twice larger memory

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                            f'#SBATCH -J {target_dir.split("/")[-2:][1]}_coincidence_{nsb}\n',
+                            f"#SBATCH --array=0-{process_size}\n",
+                            "#SBATCH --mem=30g\n",
+                            "#SBATCH -N 1\n\n",
```
why -N 1? this requires the whole machine just for this job, but it should be enough to run it just on a single task with -n 1, this way also with the smaller memory requirement we could run ~32 jobs at the same time on one machine instead of one

Elisa-Visentin replied:
-N 1 now is equivalent to -n 1 (unless you put --exclusive) but we will change it just to be sure in case something will be changed in SLURM options

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                            "ulimit -l unlimited\n",
+                            "ulimit -s unlimited\n",
+                            "ulimit -a\n\n",
+                            f"export INM={MAGIC_DL1_dir}/Merged/Merged_{str(Y_M).zfill(4)}_{str(M_M).zfill(2)}_{str(D_M).zfill(2)}\n",
```
I think instead of Merged_{str(Y_M).zfill(4)}_{str(M_M).zfill(2)}_{str(D_M).zfill(2)} you can use d, right?

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+    else:
+        coincidence_DL1_dir = f"{target_dir}/DL1/Observations"
+        if not os.path.exists(f"{coincidence_DL1_dir}/Coincident"):
+            os.mkdir(f"{coincidence_DL1_dir}/Coincident")
```
os.makedirs(...)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                            "w",
+                        ) as f:
+                            f.writelines(lines)
+    else:
```
a lot of code is repeated between the if and else cases

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                    "ulimit -a\n\n",
+                    f"export INM={nightMAGIC}\n",
+                    f"export OUTPUTDIR={nightLST}\n",
+                    "SAMPLE_LIST=($(<$OUTPUTDIR/list_LST.txt))\n",
```
this is a bit dangerous, you generate the script and put it in the queue, but the script will select which file to run based on position in a txt file that might change in the meantime. But I'm not sure how to make it more secure keeping using this nice --array option

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                    "#!/bin/sh\n\n",
+                    "#SBATCH -p short\n",
+                    f"#SBATCH -J {process_name}_coincidence\n",
+                    f"#SBATCH --array=0-{process_size}%50\n",
```
why do you limit the number of jobs to 50? I would not do so and let the cluster assign the priorities. This way if the cluster is empty at a moment it can use its whole power.

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+        list_of_coincidence_scripts = np.sort(glob.glob("LST_coincident*.sh"))
+
+        for n, run in enumerate(list_of_coincidence_scripts):
+            if n == 0:
```
how about this instead of if...else
```
launch_jobs = ""
for n, run in enumerate(list_of_coincidence_scripts):
    launch_jobs += (" && " if n>0 else "") +f"coincidence{n}=$(sbatch --parsable {run})"
````

Elisa-Visentin commented:
BEFORE MERGING/TAGGING:

- [ ] test on 1 source
- [ ] test multiple sources
- [ ] test MC + data

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,374 @@
+                if len(list_of_coincidence_scripts) < 1:
+                    continue
+                for n, run in enumerate(list_of_coincidence_scripts):
+                    if n == 0:
```
same like above

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/config_general.yaml:
```diff
@@ -0,0 +1,40 @@
+    
+data_selection:
+    source_name_database: "CrabNebula"  # MUST BE THE SAME AS IN THE DATABASE; Set to null to process all sources in the given time range.
+    source_name_output: 'Crabtest'  # Name tag of your target. Used only if source_name_database != null.
```
taking into account that we want to use those scripts to process all the data, I am not sure if allowing to have different output name is a good idea, do you have any use case for this in mind? If it is just for tests I think a better solution is to have a general directory prefix where you can specify where the files will land.

Elisa-Visentin replied:
The output name can and MUST be used only in case you process one source: e.g., a user wants to process some Crab data on its own; he can use different output names for different periods or different analyses (Crab_dec_2023, Crab_jan_2024....). Same for the input name (the one in the database): used in single-user analyses, not in massive data processing

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/LSTnsb.py:
```diff
@@ -0,0 +1,180 @@
+    """
+
+    noise = []
+    denominator = 25
```
this could be passed as a parameter with default value

Elisa-Visentin replied:
done

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/LSTnsb.py:
```diff
@@ -0,0 +1,180 @@
+    if len(run_list) < denominator:
+        mod = 1
+    else:
+        mod = int(len(run_list) / denominator)
```
instead of casting to int you can use directly integer division: 
```mod = len(run_list) // denominator```

Elisa-Visentin replied:
Thanks (C++ casting habit vs. Python :facepalm: )

Elisa-Visentin replied:
Fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/LSTnsb.py:
```diff
@@ -0,0 +1,180 @@
+    failed = 0
+    for ii in range(0, len(run_list)):
+        subrun = run_list[ii].split(".")[-2]
+        if mod == 0:
```
I think this will never trigger because you set mod to 1 if run_list has less entries then denominator. You can remove this if.

Elisa-Visentin replied:
We decrease mod while running (in case a subrun cannot be processed), so it could become 0 in case len(run_list)<(denominator+failed), where failed can increase while we run. We will have to check a bit this (just to be sure not to try a division by 0)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/LSTnsb.py:
```diff
@@ -0,0 +1,180 @@
+            except IndexError:
+                failed = failed + 1
+                if len(run_list) > denominator:
+                    mod = int(len(run_list) / (denominator + failed))
```
since you modify 'mod' in the middle of processing over the files you will probably not end up with exactly denominator entries, 
Maybe something like ```mod = (len(run_list) - ii) // (denominator -len(noise))``` would work better, but either way it is not a big problem.

Elisa-Visentin replied:
Changed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/LSTnsb.py:
```diff
@@ -0,0 +1,180 @@
+
+        if len(LST_files) > 1:
+            logger.info(
+                f"Run {run_number} classified in more than one NSB bin. Removing all these files and evaluating it again"
```
why this can happen? it is just if the old files are left?

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/create_LST_table.py:
```diff
@@ -0,0 +1,61 @@
+"""
+Create a new h5 table from the one of joint observations.
+
+Only the columns needed to produce the lists of LST runs to be processed are presenved, and two columns are added to store NSB level and error codes
```
preserved

Elisa-Visentin replied:
done

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/create_LST_table.py:
```diff
@@ -0,0 +1,61 @@
+
+    df_cut["lstchain_0.10"] = np.zeros(len(df_cut), dtype=bool)
+
+    df_cut["error_code"] = np.repeat(np.nan, len(df_cut))
```
as we discussed the error codes we can specify for individual programs

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/create_LST_table.py:
```diff
@@ -0,0 +1,61 @@
+
+    df_cut["lstchain_0.9"] = np.zeros(len(df_cut), dtype=bool)
+
+    df_cut["lstchain_0.10"] = np.zeros(len(df_cut), dtype=bool)
```
as we discussed in the call it is probably better to specify the path to the most up to date file, and to the one that was actually used in the processing

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/nsb_level.py:
```diff
@@ -0,0 +1,129 @@
+logger.setLevel(logging.INFO)
+
+
+def bash_scripts(run, date, config, env_name):
```
here we should also pass the cluster type

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/nsb_level.py:
```diff
@@ -0,0 +1,129 @@
+        "#!/bin/sh\n\n",
+        "#SBATCH -p short,long\n",
+        "#SBATCH -J nsb\n",
+        "#SBATCH -N 1\n\n",
```
-n 1

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/nsb_level.py:
```diff
@@ -0,0 +1,129 @@
+        return
+    for n, run in enumerate(list_of_bash_scripts):
+        if n == 0:
+            launch_jobs = f"nsb{n}=$(sbatch --parsable {run})"
```
check my comment on a similar look in earlier file

Elisa-Visentin replied:
done

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/nsb_to_h5.py:
```diff
@@ -0,0 +1,89 @@
+"""
+Bash scripts to run LSTnsb.py on all the LST runs by using parallel jobs
+
+Usage: python nsb_level.py (-c config.yaml)
```
the description is a copy paste error from nsb_level.py

Elisa-Visentin replied:
Fixed, Sorry!

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/nsb_to_h5.py:
```diff
@@ -0,0 +1,89 @@
+            nsb = line_str.split(",")[2]
+        df_LST = df_LST.set_index("LST1_run")
+        df_LST.loc[run, "nsb"] = float(nsb)
+        df_LST = df_LST.reset_index()
```
do you need to reset index after each line that is read? this can be pretty slow if the database grown a lot

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+        f.writelines(lines)
+
+
+def bash_stereo(target_dir, nsb, source, env_name, NSB_match):
```
pass the cluster type as one of the parameters

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+
+        for nightLST in listOfNightsLST:
+            stereoDir = f"{target_dir}/DL1/Observations/Coincident_stereo/{nightLST.split('/')[-1]}"
+            if not os.path.exists(stereoDir):
```
os.makedirs

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+            f"#SBATCH -J {process_name}_stereo\n",
+            f"#SBATCH --array=0-{process_size}%100\n",
+            "#SBATCH --mem=30g\n",
+            "#SBATCH -N 1\n\n",
```
-n 1

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+                    "#SBATCH -p short\n",
+                    f"#SBATCH -J {process_name}_stereo\n",
+                    f"#SBATCH --array=0-{process_size}%100\n",
+                    "#SBATCH -N 1\n\n",
```
-n 1

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+                    "#SBATCH -p short\n",
+                    f"#SBATCH -J {process_name}_stereo_{nsb}\n",
+                    f"#SBATCH --array=0-{process_size}\n",
+                    "#SBATCH -N 1\n\n",
```
-n 1

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+    target_dir : str
+        Path to the working directory
+    identification : str
+        Particle name. Options: protons, gammadiffuse
```
why no wobble gammas?

Elisa-Visentin replied:
Forgotten in the doc. Sorry!

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+            if n == 0:
+                launch_jobs = f"stereo{n}=$(sbatch --parsable {run})"
+            else:
+                launch_jobs = f"{launch_jobs} && stereo{n}=$(sbatch --parsable --dependency=afterany:$stereo{n-1} {run})"
```
why do you have a dependency here on the previous job? shouldn't they be independent as they are simply different runs to be processed?

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,364 @@
+                if n == 0:
+                    launch_jobs = f"stereo{n}=$(sbatch --parsable {run})"
+                else:
+                    launch_jobs = f"{launch_jobs} && stereo{n}=$(sbatch --parsable --dependency=afterany:$stereo{n-1} {run})"
```
like for MC: why there is a dependency?

Elisa-Visentin replied:
removed

Elisa-Visentin commented:
I started to fix a bit SLURM jobs (nodes and memory). Memory will then have to be fine-tuned in a 'test session' of a set of runs and sources

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+
+    for i in tqdm(range(len(list_of_nodes)), desc="Cleaning failed runs"):
+        os.chdir(list_of_nodes[i])
+        os.system('find . -type f -name "*.h5" -size -1k -delete')
```
this is potentially dangerous, I wondered what will happen if the chdir command fails (e.g. due to wrong directory path), if it will not start deleting files in a wrong directory, but it seems not to be the case, because it throws an error and stops the script. Nevertheless, one have to be very careful with autodeletes in scripts, and also there should be some reason why those files failed - they should be caught up earlier and possibly rerun.

jsitarek replied:
ok, I also did some better fitting of the name of the files to be deleted in an earlier script so we can do it also here. About failed files, still I would like to understand what happens there, there must be always some reason why it fails and at which stage the file is broken.

Elisa-Visentin replied:
I think that the issue is related to the 'base' MCP script: we open the file with the HDF5TableWriter, then we loop on the events. If one of the events fails and make the code crash (e.g., we have some missing tables in the input file), maybe the output file is produced but not filled, and we get to a very small and 'broken' output. But I am not sure about this

Elisa-Visentin replied:
After finalizing the scripts I will try to implement a few scripts/functions to update our databases with the error codes for the different steps; this way we should be able to manually remove failed jobs before moving to the next step and we could (probably) remove these automatic deletions. As for single users, in case of broken files the merging will fail and your job accounting will signal this: so they can then remove failed files

Elisa-Visentin replied:
I removed it (commented now, then we can remove it forever) for real data (we can check failed runs by job accounting) and improved the constraint on file name for MCs (we have no job monitoring for them, and we don't need it since we want to remove them in the final version)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+    ----------
+    list_of_nodes : array of str
+        List of nodes where the function will look for failed runs.
+    cwd : Path
```
since cwd is meant to return to the previous directory after changing it, wouldn't it be simpler to os.getcwd() directly in the function, instead of passing it as a variable?

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+def split_train_test(target_dir, train_fraction, source_name):
+
+    """
+    This function splits the MC proton sample in 2, i.e. the "test" and the "train" subsamples.
```
this is not something that regular analyzer would do, normally you use all the protons for training.
This is only needed if you make performance MC-only studies. Maybe worth to add this in the description of the function?

Elisa-Visentin replied:
added

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+    """
+    This function splits the MC proton sample in 2, i.e. the "test" and the "train" subsamples.
+    It generates 2 subdirectories in the directory .../DL1/MC/protons named "test" and "train" and creates sub-sub-directories with the names of all nodes.
+    For each node sub-sub-directory we move 80% of the .h5 files (if it is in the "test" subdirectory) or 20% of the .h5 files (if it is in the "train" subdirectory).
```
80% ==> 1-train_fraction 
20% ==> train_fraction

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+        Path to the working directory
+    train_fraction : float
+        Fraction of proton MC files to be used in the training RF dataset
+    source_name : str
```
why do you add here source_name (that is later on added to the path?
you might have separate MCs for different data sets, but this depends on the date (analysis period) and on NSB level), and not on actual source. Namely you can have the same source observed in dark time and mild moon with different settings, and they would end up in the same directory. On the other hand a bunch of extragal sources from the same period observed in dark time can share the same MCs.

Elisa-Visentin replied:
I will try to move the MCs out of the source directories: we can use these scripts to run a single MC production ( 1 NSB, 1 period, 1 declination...) so we can make it source independent

Elisa-Visentin replied:
And, as already discussed, this MC processing will disappear before releasing the new version of the pipeline

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+    proton_dir = f"{target_dir}/{source_name}/DL1/MC/protons"
+
+    if not os.path.exists(f"{proton_dir}/train"):
+        os.mkdir(f"{proton_dir}/train")
```
os.makedirs (without ifs)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+        range(len(list_of_dir))
+    ):  # tqdm allows us to print a progessbar in the terminal
+        if not os.path.exists(
+            f"{proton_dir}/train/{list_of_dir[directory].split('/')[-2]}"
```
since you are using all the time list_of_dir[directory].split('/')[-2] you can put it in a new variable

like above you can remove if statements if you use os.makedirs instead of os.mkdir and this way you can create all the needed subdirs on the way, i.e. you do not have to create .../train if you execute os.makedirs("../train/node_...") - this will simplify this part of code considerably

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+        list_of_runs = np.sort(
+            glob.glob(f'{proton_dir}/{list_of_dir[directory].split("/")[-2]}/*.h5')
+        )
+        split_percent = int(len(list_of_runs) * train_fraction)
```
actually it is not a percent, but a simple number (number of runs to be moved), so the name of the variable is a bit misleading

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+            f"{MAGIC_DL1_dir}/M2"
+        ):
+            if not os.path.exists(f"{MAGIC_DL1_dir}/Merged"):
+                os.mkdir(f"{MAGIC_DL1_dir}/Merged")
```
os.makedirs

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+                if os.path.exists(f"{MAGIC_DL1_dir}/M1"):
+                    for i in MAGIC_runs:
+                        if not os.path.exists(f"{MAGIC_DL1_dir}/Merged/{i[0]}"):
+                            os.mkdir(
```
os.makedirs will allow you to skip if, and also the previous os.mkdir on the parent directory

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+                            )  # Creating a merged directory for the respective night
+                        if not os.path.exists(f"{MAGIC_DL1_dir}/Merged/{i[0]}/{i[1]}"):
+                            os.mkdir(
+                                f"{MAGIC_DL1_dir}/Merged/{i[0]}/{i[1]}"
```
os.makedirs will allow you to skip if, and also the previous two os.mkdir on the two parent directories

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+                                f"{MAGIC_DL1_dir}/Merged/{i[0]}/{i[1]}"
+                            )  # Creating a merged directory for the respective run
+                        os.system(
+                            f'find  {MAGIC_DL1_dir}/M1/{i[0]}/{i[1]} -type f -name "*.h5" -size -3k -delete'
```
again, I do not like it much. If you have files that failed we should figure out why and catch it earlier

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+                            )  # Creating a merged directory for the respective night
+                        if not os.path.exists(f"{MAGIC_DL1_dir}/Merged/{i[0]}/{i[1]}"):
+                            os.mkdir(
+                                f"{MAGIC_DL1_dir}/Merged/{i[0]}/{i[1]}"
```
same as in M1, you can skip many mkdir and if statements if you use os.makedirs

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+
+    else:
+
+        process_name = f"merging_{source}"
```
you had it already defined like this before the if

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+        Tells which batch to create. Options: protons, gammadiffuse
+    env_name : str
+        Name of the environment
+    cwd : Path
```
I guess you do not have to pass this cwd as a parameter just getcwd before changing the file (unless there is some mess up of the directories when moving between the data and MCs)

Elisa-Visentin replied:
Maybe fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+    target_dir : str
+        Path to the working directory
+    identification : str
+        Tells which batch to create. Options: protons, gammadiffuse
```
no ringwobble test gammas?

Elisa-Visentin replied:
here we merge only the training samples

jsitarek replied:
closing because either way we will probably remove the MC part

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,571 @@
+def main():
+
+    """
+    Here we read the config_general.yaml file, split the pronton sample into "test" and "train", and merge the MAGIC files.
```
proton

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+"manager" that organizes the analysis process by:
+1) Creating the necessary directories and subdirectories.
+2) Generating all the bash script files that convert the
+MAGIC files from DL0 to DL1.
```
MAGIC _Y_ files are not exactly DL0, because they are already calibrated, maybe write "from _Y_ MAGIC format to DL"

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+        Telescope IDs
+    target_dir : path
+        Directory to store the results
+    noise_value : list
```
I do not understand this part. The script is meant only for calib to DL1 files of the data, so why do you have the part of adding the noise, that is only done for MCs?

Elisa-Visentin replied:
To 'update' the standard config (written by this function) with the values we want to use to process the MCs (extra_noise_in_dim_pixels...)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+):
+
+    """
+    Below we create a bash script that links the the MAGIC data paths to each subdirectory.
```
the the ==> the

Elisa-Visentin replied:
fixed

jsitarek replied:
I still see "the the MAGIC data path" in the line above

Elisa-Visentin replied:
Because I did not push the commit :woman_facepalming:

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+                        f"ls $IN1/*{i[1][-2:]}.*_Y_*.root > $OUT1/list_dl0.txt\n",
+                    ]
+                    f.writelines(lines)
+        else:
```
the lines for NSB_MATCH False look the same as above, or am I missing something?
they can be extracted from the if to make the code shorter

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+            os.mkdir(f"{target_dir}/v{__version__}/{source_name}")
+        if not os.path.exists(f"{target_dir}/v{__version__}/{source_name}/DL1"):
+            os.mkdir(f"{target_dir}/v{__version__}/{source_name}/DL1")
+        dl1_dir = str(f"{target_dir}/v{__version__}/{source_name}/DL1")
```
dl1_dir = ...
os.makedirs (dl1_dir) 
instead of the ifs above

Elisa-Visentin replied:
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+            os.mkdir(f"{target_dir}/v{__version__}/{source_name}/DL1")
+        dl1_dir = str(f"{target_dir}/v{__version__}/{source_name}/DL1")
+    else:
+        if not os.path.exists(target_dir):
```
again you can use os.makedirs
the current code will fail if target_dir was already created, but it did not have DL1 inside it

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/setting_up_config_and_dir.py:
```diff
@@ -0,0 +1,740 @@
+            if (args.analysis_type == "onlyMC") or (
+                args.analysis_type == "doEverything"
+            ):
+                lists_and_bash_generator(
```
you can do it simpler with a loop
```
MCs=[("gammas", MC_gammas), ("electrons", MC_electrons), ...]
for MCtype, MCdir in MCs:
    lists_and_bash_generator( MCtype, targetdir, MCdir, ...)
```

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/list_from_h5.py:
```diff
@@ -0,0 +1,275 @@
+    )  # TODO: put this file in a shared folder
+
+    df_LST = split_lst_date(df_LST)
+    df_LST = df_LST.astype(
```
I tried to run it and got an error because of NANs in LST1_run

Elisa-Visentin replied:
Now it should be fixed

aleberti commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,234 @@
+
+Authorized institute server (Client) &rarr;  ssh connection to CTALaPalma &rarr; ssh connection to cp01/02.
+
+2) Once connected to the IT Container, install magic-cta-pipe (e.g. in your home directory in the IT Container) with the following commands (if you have mamba installed, we recommend you to use it instead of conda. The installation process will be much faster.):
```
here you assume that the user has already anaconda/miniconda/miniforge. I would add that those should be better installed in the workspace directory i.e. `/fefs/aswg/workspace/yourname`. Cloning only MCP into the home directory is then fine.

Elisa-Visentin replied:
added

aleberti commented on magicctapipe/scripts/lst1_magic/README.md:
````diff
@@ -0,0 +1,234 @@
+Finding LST runs...
+Finding MAGIC runs...
+```
+And it will save the files TARGET_LST_runs.txt, TARGET_MAGIC_runs.txt, and list_sources.dat in your working directory.
````
what if no runs are found? are the files created anyway but with nothing in them? or will the `list_from_h5` script say something about that?

Elisa-Visentin replied:
It should not create the files. I will check this and add a warning if needed

Elisa-Visentin replied:
I added a return in the script in case no run can be found. I wlll add also a note in the readme

aleberti commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,234 @@
+
+Since the DL3 may have only a few MBs, it is typically convenient to download it to your own computer at this point. It will be necessary to have astropy and gammapy (version > 0.20) installed before proceeding. 
+
+We prepared a [Jupyter Notebook](https://github.com/ranieremenezes/magic-cta-pipe/blob/master/magicctapipe/scripts/lst1_magic/SED_and_LC_from_DL3.ipynb) that quickly creates a counts map, a significance curve, an SED, and a light curve. You can give it a try. 
```
why not committing the notebook to the main repository? (into the notebooks driectory, and not the scripts one)

Elisa-Visentin replied:
actually this notebook does not seem to exist in the current branch. We only have to remove this sentence from the readme

Elisa-Visentin replied:
removed

aleberti commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/clusters.py:
```diff
@@ -0,0 +1,41 @@
+"""
+
+
+def slurm_lines(p, J, array=None, mem=None, out_err=None):
```
variable names here could be a bit different and more descriptive:

- p --> queue
- J --> job_name
- out_err --> out_name (since it is used for both out and err, just with different extension)

jsitarek replied:
I thought first to use such descriptive names, but since either way we describe them in the doc string of the function I used directly the names that are used in slurm. But if you prefer I can change them to queue, job_name and out_name

aleberti replied:
I would prefer if they are changed :) even if they are function parameters, it is still better to use descriptive names

jsitarek replied:
ok, done

aleberti commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/clusters.py:
```diff
@@ -0,0 +1,41 @@
+"""
+Module for generating bash script lines for running analysis in different clusters
+"""
+
```
I think here you should also add:

```python
__all__ =  ["slurm_lines"]
```

jsitarek replied:
ok, I will add it, but note that it does not change anything, because by default  all functions are exported

aleberti commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/clusters.py:
```diff
@@ -0,0 +1,41 @@
+    Returns
+    -------
+    list
+        List of strings
```
I would say it is "List of strings to submit a SLURM job."

aleberti commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/clusters.py:
```diff
@@ -0,0 +1,41 @@
+    array : None or int
+        If not none array of jobs from 0 to array will be made
+    mem : None or str
+        Requested memory
```
I would add `If None, no specific memory will be requested`. Also, may an example of what can be specified e.g. "16g".

jsitarek replied:
this is actually no true, if you do not specify the this parameter the default of the cluster setting is used, which I think is 5 GB

Elisa-Visentin commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/database_production/create_LST_table.py:
```diff
@@ -0,0 +1,61 @@
+            key="joint_obs",
+        )
+        df_cut = pd.concat([df_old, df_cut]).drop_duplicates(
+            subset="LST1_run", keep="first"
```
We don't use this script to see which is the newest version. We use it only to update and then we run lstchain_version over the whole dataframe to update version list and path to last file (this takes only a couple of minutes, and we don't have to check if something changed in the versions stored in the 'old' dataframe before appending). We append only not to lose error codes and NSB values while updating the table

Elisa-Visentin commented:
I marked it as ready for review so that we can check the status of unresolved conversations. We should start to close the resolved ones, just not to get lost :sweat_smile:

jsitarek commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,226 @@
+
+We prepared a [Jupyter Notebook](https://github.com/ranieremenezes/magic-cta-pipe/blob/master/magicctapipe/scripts/lst1_magic/SED_and_LC_from_DL3.ipynb) that quickly creates a counts map, a significance curve, an SED, and a light curve. You can give it a try. 
+
+The folder [Notebooks](https://github.com/cta-observatory/magic-cta-pipe/tree/master/notebooks) contains Jupyter notebooks to perform checks on the IRF, to produce theta2 plots and SEDs. Note that the notebooks run with gammapy v0.20 or higher, while the gammapy version adopted in the MAGIC+LST-1 pipeline is v0.19.
```
I removed it myself. I hope you do not mind :-)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,262 @@
+
+        MAGIC_DL1_dir = f"{target_dir}/v{__version__}/{source_name}/DL1"
+    else:
+        coincidence_DL1_dir = f"{target_dir}/{source_name}/DL1/Observations"
```
note that now we have /v{__version__} also in NSB_match case

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,262 @@
+        file_name = f"{target_dir}/{source_name}/config_coincidence.yaml"
+    else:
+        file_name = f"{target_dir}/v{__version__}/{source_name}/config_coincidence.yaml"
+    with open(file_name, "w") as f:
```
as in the other file: this creating of config files can be improved: we can use directly the config file of the repository, or pass a path to the processing config file in the config_general.yaml file of the semi automatic scripts and use this instead

Elisa-Visentin replied:
We need the coincidence 'parameters' from the config.yaml in the repo and the tel. IDs from the general one. And both infos must be passed to the coincidence script that we call in conda

jsitarek replied:
yes, but why not include the path to config file (the normal one of MCP) as a parameter in config_general (the specific config file for automatic processing)?

Elisa-Visentin replied:
The one we have in the 'resource' folder? It is in a fixed (and a bit 'hidden') path, so I don't know if every user could be able to provide its path. We could allow for a default path (the one in conda/envs/....) to be used if no path is provided in the config_general

jsitarek replied:
yes, I am not sure if there is a simpler method, but I think you can do it e.g.:
```
import magicctapipe
default_path = magicctapipe.__file__.replace('/__init__.py', '/resources/config.yaml')
```

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,262 @@
+                    "SAMPLE_LIST=($(<$OUTPUTDIR/logs/list_LST.txt))\n",
+                    "SAMPLE=${SAMPLE_LIST[${SLURM_ARRAY_TASK_ID}]}\n",
+                    "export LOG=$OUTPUTDIR/logs/coincidence_${SLURM_ARRAY_JOB_ID}_${SLURM_ARRAY_TASK_ID}.log\n",
+                    f"conda run -n {env_name} lst1_magic_event_coincidence --input-file-lst $SAMPLE --input-dir-magic $INM --output-dir $OUTPUTDIR --config-file {target_dir}/v{__version__}/{source_name}/config_coincidence.yaml >$LOG 2>&1",
```
this will also need a similar treatment with storing error codes. I can do it.

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,258 @@
+
+
+def linking_bash_lst(
+    target_dir, LST_runs, source_name, LST_version, env_name, NSB_match
```
including again the comment because it got outdated due to a different change in the same line:
"you should pass also queueing system type to this function because you produce a bash file following slurm style"

Elisa-Visentin replied:
added

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,302 @@
+    else:
+        file_name = f"{target_dir}/v{__version__}/{source_name}/config_stereo.yaml"
+    with open(file_name, "w") as f:
+        lines = [
```
adding old comment that got outdated because of a different change in the same line:
"similar like in the other file, it would be better to extract those lines from the config file"

Elisa-Visentin replied:
done

jsitarek replied:
but you do not really extract it from the file.
wouldn't something like ```lines = {mc_tel_ids: config_dict['mc_tel_ids']}``` work instead?

Elisa-Visentin replied:
fixed (in the three scripts)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/stereo_events.py:
```diff
@@ -0,0 +1,302 @@
+        yaml.dump(conf, f, default_flow_style=False)
+
+
+def bash_stereo(target_dir, source, env_name, NSB_match):
```
old comment: 
"pass the cluster type as one of the parameters"

Elisa-Visentin replied:
Done

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,413 @@
+        range(len(list_of_dir))
+    ):  # tqdm allows us to print a progessbar in the terminal
+
+        os.makedirs(f"{proton_dir}/train/{list_of_dir[directory].split('/')[-2]}")
```
re-adding old comment:
"since you are using all the time list_of_dir[directory].split('/')[-2] you can put it in a new variable"

Elisa-Visentin replied:
done

Elisa-Visentin commented:
Fixed lstchain config  (added an option to solve run-time errors for some of the runs)  and MC file in NSB evaluation. Added cluster name. Simplified a bit the code. Minor fixes

Elisa-Visentin commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,226 @@
+
+1) The very first step to reduce MAGIC-LST data is to have remote access/credentials to the IT Container, so provide one. Once you have it, the connection steps are the following:
+
+Authorized institute server (Client) &rarr;  ssh connection to CTALaPalma &rarr; ssh connection to cp01/02.
```
done

Elisa-Visentin commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,234 @@
+MAGIC+LST analysis starts from MAGIC calibrated data (\_Y\_ files), LST  data level 1 (DL1) data and SimTelArray DL0 data, and our goal is to achieve data level 3 (DL3).
+
+Behind the scenes, the semi-automatic scripts will run:
+- `magic_calib_to_dl1` on real MAGIC data, to convert them into DL1 format.
```
fixed

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,275 @@
+    env_name : str
+        Name of the conda environment
+    NSB_match : bool
+        If real data are matched to pre-processed MCs or not
```
cluster is missing in the doc string

Elisa-Visentin replied:
added

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,275 @@
+        MAGIC_DL1_dir = f"{target_dir}/v{__version__}/{source_name}/DL1/Observations/"
+
+    dates = [os.path.basename(x) for x in glob.glob(f"{MAGIC_DL1_dir}/Merged/Merged_*")]
+    if cluster == 'SLURM':
```
for the moment 'SLURM' is the only option, but when we introduce other clusters a lot of code can be shared between them, so we should not have if cluster == AAA < long code> elif cluster == BBB <similar long code>.
but rather set proper variables (like paths) before the bulk of the code. 
Since it is difficult to know now which variables will need to be set how about you make a simple if cluster != 'SLURM' <exit with an error message> and have the rest of the code one indent less complicated :-)

Elisa-Visentin replied:
Done

jsitarek commented on magicctapipe/scripts/lst1_magic/README.md:
```diff
@@ -0,0 +1,223 @@
+
+1) The very first step to reduce MAGIC-LST data is to have remote access/credentials to the IT Container. If you do not have it, please write an email to request it to <admin-ctan@cta-observatory.org>, and the admin will send you the instructions to connect to the IT container.
+
+2) Once connected to the IT Container, install magic-cta-pipe (e.g. in your home directory in the IT Container) with the following commands (if you have mamba installed, we recommend you to use it instead of conda, so that the installation process will be much faster; if you don't have anaconda/miniconda/miniforge, please install one of them into your workspace directory):
```
now that I think about this, this installation explanation is  kind of repeating what is said about installation in the general Readme file that is accessible from the main repository (just recommending mamba instead of conda). Since it is more natural that users see the magicctapipe/scripts/lst1_magic/README.md file only after installation, I would recommend to remove the installation descriptions from here, and instal update the main README file.

Elisa-Visentin replied:
Removed

jsitarek commented on magicctapipe/resources/lstchain_standard_config_modified.json:
```diff
@@ -0,0 +1,289 @@
+{
```
why do you need the standard LST config file?? LST chain gets installed together with MCP, so you can take it from there

Elisa-Visentin replied:
We need the modified one to evaluate NSB for some particular runs (by adding the heuristic_flatfield lin, not present in the config installed in lstchain when installing MCP, some runs can be processed instead of producing a 'missing FF table' error)

Elisa-Visentin replied:
I can try to find out which are the needed entries to create the dictionary and modify it. But as far as I know I will have to produce a new file instead of passing the dictionary to the function (but I have to check lstchain)

Elisa-Visentin replied:
maybe fixed now (I load and modify the config in nsb_level and I save the (un)changed version to the current directory to use it in LSTnsb)

joannaWojS commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/config_general.yaml:
```diff
@@ -0,0 +1,42 @@
+    
+general:
+    SimTel_version: "v1.4"  
+    LST_version   : "v0.10"
```
I've noticed that on my computer, the script list_from_h5 works correctly only with version 0.9 LST_version in config file. When I switch to version 0.10, the script does not seem to locate the LST runs. I suspect the issue might be related to the observations_LST.h5 file, specifically the lstchain-version noted in the 8th column, which remains set to v0.9.

Elisa-Visentin replied:
Thank you. I will check the LST database (there were some bugs in the related scripts and maybe I still get the wrong version). For sure the list_from_h5 scripts check if the version requested in the config_general is the same of the processed file for NSB (i.e., generally the last available version for a run), so in case of bugs in file versions the files cannot be found. I will check it hopefully today/tomorrow

joannaWojS replied:
I didn't use config_general with v0.10. I have just taken another look at your table and you're right that some rows are labeled v0.10 and others v0.9. I worked with data from March 2022, and it took me some time to figure out why list_from_h5 couldn't find the LST runs. Therefore, I think we should add a brief comment in the config file reminding users to check the LST version.

Elisa-Visentin replied:
I will add it (also in the Readme). Thanks

Elisa-Visentin replied:
Actually, the script that writes the lists check for the lstchain_processed_file version (i.e., last version), i.e., the one used to evaluate the NSB: in case of no errors, the NSB value should be the same for 0.9 and 0.10 if both are present and it is always better to use the last available version; in case of errors (e.g., missing flatfield), we discard the run if the NSB evaluation fails and we don't propagate the errors to the other steps.
So, you list the run with 0.10 version if you use 0.10 in the config AND if the file processed to evaluate the NSB is the 0.10 one

Elisa-Visentin commented:
- [ ] doc + readme
- [ ] update database with error codes

Elisa-Visentin commented:
I tried to make the directory structure a bit simpler (also for job accounting): Since now MC are out of the source folders there should be no need to use the 'Observations' subfolders when we process both data and MC. I will fix now the CI and start an extensive test 8data + MC) to check if everything is fine

Elisa-Visentin commented:
TODO: PR #229 , #238, #239 , #234

Elisa-Visentin commented:
job accounting fine also for Coincidence and Stereo (no changes needed)!

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/merging_runs.py:
```diff
@@ -0,0 +1,362 @@
+    cwd = os.getcwd()
+    for i in tqdm(range(len(list_of_nodes)), desc="Cleaning failed runs"):
+        os.chdir(list_of_nodes[i])
+        os.system('find . -type f -name "dl1_[gphe]*_zd*_az*.h5" -size -1k -delete')
```
we were discussing that automatic removal is dangerous and we should rather understand the reason for files failing at earlier step, and @Elisa-Visentin wrote that she removed (commented out) this, but I still see that the removal is being done. Was is not commented out after or, or it turned out to be needed for some other reason?

Elisa-Visentin replied:
I removed it from real data. I can remove it also from MC and add a warning in the readme (i.e.,' if you are processing MCs, check Dl1s for failed jobs and remove broken files before merging') Broken files must be removed not to make the merging jobs fail, but if you process MCs you must know what you are doing (and know the pipeline!) so you can take care of manually removing files

jsitarek replied:
since the MC part will be removed, there is no point in modifying this part now.
I am resolving this comment

Elisa-Visentin replied:
BTW, this part was removed in commit 3c47546 a couple of weeks ago (then probably this change got lost in the 'Outdated' comments)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,253 @@
+                    queue="short",
+                    job_name=f"{source_name}_coincidence",
+                    array=process_size,
+                    mem="8g",
```
from the tests done by @joannaWojS coincidence script was using on average 1 G of memory and the largest files were 3.4G, so I think we can lower it a bit, how about 4.5g? Or did you get any files where more memory than this was required?

Elisa-Visentin replied:
On a test I got to a maximum about 5 g, so maybe we can reduce it to 6 but I won't reduce it to 4.5 or some jobs will fail and will have to be launched again

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/coincident_events.py:
```diff
@@ -0,0 +1,253 @@
+        source_list = joblib.load("list_sources.dat")
+
+    else:
+        source_list.append(source)
```
shouldn't it be ```source_list.append(source_in)``` ?
however then ```source``` would not be used at all

the two source names are a bit confusing

Elisa-Visentin replied:
source_in is the name in the database, source is the 'tag' used for the source (and the name of the folder in which you store the results): e.g., source_in=CrabNebula, source=Crab_23

jsitarek replied:
but the check is if source is not None
so it might happen that somebody puts source_in=CrabNebula, but does not put source, and then you add None to the list. 
You could add a condition that if source is None you put source = source_in

Elisa-Visentin replied:
I will add it now in a small PR (the same 'structure' is used in many scripts)

jsitarek commented on magicctapipe/scripts/lst1_magic/semi_automatic_scripts/config_general.yaml:
```diff
@@ -0,0 +1,40 @@
+mc_tel_ids:
```
mc_tel_ids are also in resources/config.yaml

also we could config_general.yaml to config_autoMCP.yaml or something like this, because "general" sounds more like something that is used for all the scripts not only the semi-automatic ones

Elisa-Visentin replied:
the config.yaml gets installed when we install the pipeline: so, if we want to change the tels:id it is much easier if we use the config_general (I agree, I will change the name) because we don't need a new 'pip install' to actually change the ids.
If you don't use the automatic scripts you have to copy somewhere the config.yaml, modify it and pass the modified version to the scripts; if you use the automatic ones and you don't want to change cleaning.... you can just modify the config general and don't care about what is installed in the pipeline

jsitarek replied:
modified config file you can pass to the config_auto_MCP:(general => base_config_file)
and if you want to modify the telescopes you need to either way extract and modify one of the config files. 
The main problem is that if we have mc_tel_id in both configs it is very confusing which one is really used where

Elisa-Visentin replied:
We rarely modify the IDs, just in case we have to run special analyses, so i think that we can try to remove the IDs from the config_auto_MCP and check if it is user friendly. As above, new PR :)   (hopefully today)

Elisa-Visentin commented:
TODO: PR https://github.com/cta-observatory/magic-cta-pipe/pull/240
check console scripts
check database paths (config)
test real data
readme

Elisa-Visentin commented:
@joannaWojS and @jsitarek I just fixed the check_MAGIC_runs by allowing it to read dates from user and database paths from the config (just a copy-paste of the code from #240)

Elisa-Visentin commented:
The readme should be fine. We only miss the last PR and the last final quick checks 🥳

Elisa-Visentin commented:
Recap:
- [ ] console scripts
- [ ] db paths
- [ ] nsb (first results seem promising...)
- [ ] remove MC
- [ ] minor refactoring update magic db
- [ ] open conversations
Anything else for the last final version?
